### PR TITLE
Load report to s3 and SES

### LINKED ISSUE #31: report: create report script
### Task

- [x] create a script that generates a summary report based on the data shown in the dashboard
- [x] structure the pdf in a digestible and appealing manner
- [x] save this pdf in an S3 bucket

### User Story

as a developer, this script is necessary to ensure users receive their summaries

### LINKED ISSUE #75: Report: Load PDF to bucket and SES
### Task

- [x] Send PDF report to S3 bucket
- [x] Send PDF to subscribers

### User Story

As a user, I want to be able to receive the PDF reports so I can make informed decisions

### LINKED ISSUE #99: Test load PDF
### Task

- [x] Test loading to S3 bucket
- [x] Test formatting and sending email

### User Story
As a developer, I want to test that the functionality of loading the PDF works as expected and reduces the chances of bugs to occur in the future.

### PR DESCRIPTION
**Pull request by FarihaChoudhury**
Loads PDF report into an S3 bucket and sends to all verified subscribers on the subscribers list. 
Tests loading functionality.

Closes #75, closes #99, closes #31.

### CODE REVIEW

File: pdf_report/README.md [hunk 1]
```diff
@@ -1,2 +1,21 @@
...
+```text
+AWS_ACCESS_KEY=your_aws_access_key
+AWS_SECRET_KEY=your_aws_secret_access_key
+DB_PASSWORD=your_database_password
+DB_USERNAME=your_database_username
+DB_NAME=your_database_name
+DB_PORT=your_database_port
+DB_IP=your_database_ip_or_hostname
+SOURCE_EMAIL=your_email_to_send_ses
+S3_BUCKET_NAME=your_s3_bucket_name
```
Comment (A-Faris) [inline pdf_report/README.md]:
```suggestion
AWS_ACCESS_KEY=your_aws_access_key
AWS_SECRET_KEY=your_aws_secret_access_key

DB_PASSWORD=your_database_password
DB_USERNAME=your_database_username
DB_NAME=your_database_name
DB_PORT=your_database_port
DB_IP=your_database_ip_or_hostname

SOURCE_EMAIL=your_email_to_send_ses
S3_BUCKET_NAME=your_s3_bucket_name 
```

File: pdf_report/load_pdf.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
+""" Loads the PDF summary report of RealTimeTrains data into an S3 bucket. """
+from os import environ, path
+import logging
```
Comment (A-Faris) [inline pdf_report/load_pdf.py]:
```suggestion
""" Loads the PDF summary report of RealTimeTrains data into an S3 bucket. """

from os import environ, path
import logging
```

File: pdf_report/load_pdf.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from io import BytesIO
+from datetime import datetime
+
+from dotenv import load_dotenv
+from boto3 import client
+from botocore.exceptions import NoCredentialsError
+import botocore.exceptions
+from botocore.exceptions import ClientError
```
Comment (A-Faris) [inline pdf_report/load_pdf.py]:
```suggestion
from botocore.exceptions import NoCredentialsError, ClientError
```

File: pdf_report/load_pdf.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                           aws_secret_access_key=environ['AWS_SECRET_KEY'])
+        return s3_client
+    except NoCredentialsError:
+        logging.error("Error, no AWS credentials found")
+        return None
+
+
+def upload_pdf_data_to_s3(s3: client, bucket_name: str, s3_filename: str, pdf: BytesIO):
+    """ Uploads pdf of summary report to S3 bucket. """
+
```
Comment (A-Faris) [inline pdf_report/load_pdf.py:39]:
```suggestion
```

File: pdf_report/load_pdf.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    except FileNotFoundError:
+        logging.error("The file was not found.")
+    except Exception:
+        logging.error(
+            "Error occurred when connecting and uploading to S3 bucket.")
+
+
+def get_connection() -> connection:
+    """ Retrieves connection and returns it. """
+
```
Comment (A-Faris) [inline pdf_report/load_pdf.py:64]:
```suggestion
```

File: pdf_report/load_pdf.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        password=environ['DB_PASSWORD'],
+        host=environ['DB_IP'],
+        port=environ['DB_PORT'],
+        dbname=environ['DB_NAME']
+    )
+
+
+def get_cursor(conn: connection) -> cursor:
+    """ Retrieves cursor and returns it. """
+
```
Comment (A-Faris) [inline pdf_report/load_pdf.py:77]:
```suggestion
```

File: pdf_report/load_pdf.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    return conn.cursor(cursor_factory=RealDictCursor)
+
+
+def get_subscribers(conn: connection) -> list:
+    """ Queries the subscriber table, returning a list of all subscriber emails."""
+
```
Comment (A-Faris) [inline pdf_report/load_pdf.py:83]:
```suggestion
```

File: pdf_report/test_load_pdf.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
+""" Unit tests to test load PDF functions. """
+import unittest
+from unittest.mock import MagicMock, patch
+import os
+from io import BytesIO
+from datetime import datetime
+from email.mime.multipart import MIMEMultipart
+from botocore.exceptions import ClientError
+import psycopg2
+from psycopg2.extensions import connection, cursor
```
Comment (A-Faris) [inline pdf_report/test_load_pdf.py]:
```suggestion
""" Unit tests to test load PDF functions. """

import unittest
from unittest.mock import MagicMock, patch
from os import environ
from io import BytesIO
from datetime import datetime
from email.mime.multipart import MIMEMultipart

from botocore.exceptions import ClientError
from psycopg2.extras import RealDictCursor
from psycopg2.extensions import connection, cursor
```

File: pdf_report/test_load_pdf.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    get_subscribers,
+    create_ses_client,
+    send_email,
+    get_s3_client,
+    upload_pdf_data_to_s3
+)
+
+
+@patch.dict(
+    os.environ,
```
Comment (A-Faris) [inline pdf_report/test_load_pdf.py]:
```suggestion
    environ,
```

File: pdf_report/test_load_pdf.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    @patch("load_pdf.get_connection")
+    def test_get_cursor(self, mock_get_connection: connection):
+        """ Tests get_cursor returns cursor. """
+
+        mock_connection = MagicMock()
+        mock_get_connection.return_value = mock_connection
+        db_cursor = get_cursor(mock_connection)
+
+        mock_connection.cursor.assert_called_once_with(
+            cursor_factory=psycopg2.extras.RealDictCursor)
```
Comment (A-Faris) [inline pdf_report/test_load_pdf.py]:
```suggestion
            cursor_factory=RealDictCursor)
```

File: pdf_report/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,3 @@
+boto3
+python-dotenv
+xhtml2pdf
```
Comment (A-Faris) [inline pdf_report/requirements.txt]:
```suggestion
xhtml2pdf
psycopg2-binary
```

**Review by A-Faris — DISMISSED** [review]
Looks good

File: pdf_report/load_pdf.py (hunk 3 above)
Comment (FarihaChoudhury) [reply pdf_report/load_pdf.py]: I will fix it thanks

File: pdf_report/load_pdf.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+def get_s3_client() -> client:
+    """ Returns s3 client. """
+    load_dotenv()
+    try:
+        s3_client = client('s3',
+                           aws_access_key_id=environ['AWS_ACCESS_KEY'],
+                           aws_secret_access_key=environ['AWS_SECRET_KEY'])
+        return s3_client
```
Comment (A-Faris) [inline pdf_report/load_pdf.py]:
```suggestion
        return client('s3',
                           aws_access_key_id=environ['AWS_ACCESS_KEY'],
                           aws_secret_access_key=environ['AWS_SECRET_KEY'])
```

File: pdf_report/load_pdf.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    except NoCredentialsError:
+        logging.error("Error, no AWS credentials found")
+        return None
+
+
+def upload_pdf_data_to_s3(s3: client, bucket_name: str, s3_filename: str, pdf: BytesIO):
```
Comment (A-Faris) [inline pdf_report/load_pdf.py]:
```suggestion
def upload_pdf_data_to_s3(s3: client, bucket_name: str, s3_filename: str, pdf: BytesIO) -> None:
```

File: pdf_report/load_pdf.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    """ Uploads pdf of summary report to S3 bucket. """
+
+    s3.upload_fileobj(pdf, bucket_name, s3_filename)
+    logging.info("pdf file uploaded %s", s3_filename)
+
+
+def upload_pdf_to_s3():
```
Comment (A-Faris) [inline pdf_report/load_pdf.py]:
```suggestion
def upload_pdf_to_s3() -> None:
```

File: pdf_report/test_load_pdf.py
Comment (A-Faris) [inline pdf_report/test_load_pdf.py:1]: Nice tests

**Review by A-Faris — DISMISSED** [review]
Looks good

**Review by A-Faris — DISMISSED** [review]
Looks good

File: pdf_report/load_pdf.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+from psycopg2.extensions import connection, cursor
+from psycopg2.extras import RealDictCursor
+
+
+LOCAL_FILE_PATH = 'performance_report.pdf'
+
+
+def get_s3_client() -> client:
+    """ Returns s3 client. """
+    load_dotenv()
```
Comment (Zhi-704) [inline pdf_report/load_pdf.py]: This should be moved outside of functions unless it's the main function

File: pdf_report/load_pdf.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        logging.error("The file was not found.")
+    except Exception:
+        logging.error(
+            "Error occurred when connecting and uploading to S3 bucket.")
+
+
+def get_connection() -> connection:
+    """ Retrieves connection and returns it. """
+
+    load_dotenv()
```
Comment (Zhi-704) [inline pdf_report/load_pdf.py]: Same case here

File: pdf_report/load_pdf.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+    """ Sends the PDF of the performance report via email to subscribers
+        Loads the PDF report to the s3 bucket with timestamp."""
+    email_performance_report()
+    upload_pdf_to_s3()
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO,
+                        format="%(asctime)s - %(levelname)s - %(message)s")
+
```
Comment (Zhi-704) [inline pdf_report/load_pdf.py:171]: load_dotenv() should be here instead

File: pdf_report/load_pdf.py (hunk 16 above)
Comment (FarihaChoudhury) [reply pdf_report/load_pdf.py]: :( okay

**Review by Zhi-704 — APPROVED** [review]
Looks good!

**Pull request merged**